pkgs/applications/misc/plover/default.nix

```diff
@@ -24,32 +29,78 @@
     ];
   };
 
-  dev = with python36Packages; mkDerivationWith buildPythonPackage rec {
+  dev = with python3Packages; mkDerivationWith buildPythonPackage rec {
     pname = "plover";
-    version = "4.0.0.dev8";
+    version = "4.0.0.dev9";
 
     meta = with lib; {
       description = "OpenSteno Plover stenography software";
       maintainers = with maintainers; [ twey kovirobi ];
-      license     = licenses.gpl2;
+      license     = licenses.gpl2Plus;
     };
 
-    src = fetchurl {
-      url    = "https://github.com/openstenoproject/plover/archive/v${version}.tar.gz";
-      sha256 = "1wxkmik1zyw5gqig5r0cas5v6f5408fbnximzw610rdisqy09rxp";
+    src = fetchFromGitHub {
+      owner = "openstenoproject";
+      repo = "plover";
+      rev = "54dbcf4ea73cc1ecc1d7c70dbe7bdb13f055d101";
+      sha256 = "1jm6rajlh8nm1b1331pyvp20vxxfwi4nb8wgqlkznf0kkdvfa78a";
     };
 
-    # I'm not sure why we don't find PyQt5 here but there's a similar
-    # sed on many of the platforms Plover builds for
-    postPatch = "sed -i /PyQt5/d setup.cfg";
-
     checkInputs           = [ pytest mock ];
-    propagatedBuildInputs = [ Babel pyqt5 xlib pyserial appdirs wcwidth setuptools ];
+    propagatedBuildInputs = [
+      Babel pyqt5 xlib pyserial
+      appdirs wcwidth setuptools
+      certifi
+    ];
 
     dontWrapQtApps = true;
 
     preFixup = ''
       makeWrapperArgs+=("''${qtWrapperArgs[@]}")
     '';
   };
+
+  plugins-manager = with python3Packages; buildPythonPackage rec {
+    pname = "plover-plugins-manager";
+    version = "0.6.1";
+
+    src = fetchFromGitHub {
+      owner = "benoit-pierre";
+      repo = "plover_plugins_manager";
+      rev = version;
+      sha256 = "sha256-7OyGmSwOvoqwbBgXdfUUmwvjszUNRPlD4XyBeJ29vBg=";
+    };
+
+    patches = [ ./plugins_manager.patch ];
+
+    buildInputs = [
+      # plover.dev
+      dev
+    ];
+
+    propagatedBuildInputs = [
+      pip pkginfo pygments
+      readme_renderer requests
+      requests-cache requests-futures
+      setuptools wheel
+    ];
+
+    # tests try to instantiate a virtualenv and lack permission
+    doCheck = false;
+
+    meta = with lib; {
+      description = "OpenSteno Plover stenography software plugin manager";
+      homepage = "https://github.com/benoit-pierre/plover_plugins_manager";
+      license = licenses.gpl2Plus;
+      maintainers = with maintainers; [ evils ];
+    };
+  };
+
+  dev-with-plugins = dev.overrideAttrs (old: {
```

Comment:
I can only echo my thoughts on #110658 that included the plugin manager is expected functionality with the dev branch of plover. It's what you'll get if you're installing plover dev with any other package manager, or the appimage, or whatever else.

If anything, now that my experience with Nix has grown, I'd like to see an argument that can be overridden (to `false`) on plover dev that enables the package manager by default, rather than creating two separate derivations that are built, something like

```nix
plover.dev.override {
  withPluginManager = false;
}
```

Reply:
Actually, thinking further, I'm not exactly happy that the `plugin-manager` derivation is exposed as it is either.

While there is the potential usecase that one would want to override the plugin manager, the way it's currently exposed gives no particular ability to do so. It shouldn't be exposed in the `plover` tree's top level as it currently is. My best image for how it *should* be exposed while giving a way to override it probably looks like putting it under a subtree like `plover.devPlugins.plugin-manager`, then `plover.dev` could expose, perhaps, both a `withPluginManager` and `plugins` override argument. If you wanted to swap out the plugin manager with an overridden version, you would set `withPluginManager` to `false`, then add your overridden version to `plugins`.

I'm more than happy for more brainstorming on the subject, though, I admit this probably isn't the most elegant solution but I think we need to deal with a couple of concerns, including:

 1. the plugin-manager being included being an expected default
 2. allowing users to disable it if they wish
 3. making the `plover` subtree not bloated, and
 4. considering the possibility of extending plover with a `withPlugins`-like functionality in the future, which this is already great groundwork for

Reply:
the `right solution™` is probably to accept a list of plugins
then the (default) setting of an empty list would not provide the plugin manager

as far as i know this would require packaging all the plugins in nixpkgs
which i'm not willing to do because plover doesn't work on wayland...


for now, setting plover.dev to include the plugin manager does seem like the right way forward

unfortunately i seem to recall the issue with this PR being,
plover not reliably finding the site packages (plugins) installed by the plugin manager
(between a nix-env install, a nix-shell, and a NixOS systemPackages install)
which is something i don't know how to fix, but would like to see fixed before seeing this PR merged
alternatively, we could merge something that works in 1 use-case and let users get annoyed enough to fix it themselves

Reply:
You could probably see if it's possible to pass the site packages in an environmental variable?

It would need to be the same place recognized by the plugin manager and writeable though, but the only two cases I see are:

 1. executed by a user *with* a $HOME directory, and
 2. executed by a user *without* a $HOME directory (which, frankly, given the use-case of plover may not be a problem in the first place)

Reply:
This can be done with `makeWrapper` using the `--run` flag and a `'`-quoted argument to it using `$HOME`, you can see how my emacs config using Nix does similar [here](https://github.com/lambdadog/emacs-config/blob/bb49dd356a6c2e9c37ab0c0b1b1144f2497acb95/emacsWithConfig/default.nix#L24)

I imagine something like (assuming the variable needed is `SITE_PACKAGES` and both the plugin manager and plover would read from it correctly)

```bash
makeWrapper "PATH/TO/PLOVER/EXECUTABLE" "$out/bin/plover"
  --run 'mkdir -p "$HOME/PATH/TO/YOUR/SITEPACKAGES"'
  --run 'export SITE_PACKAGES="$HOME/PATH/TO/YOUR/SITEPACKAGES"'
```

would work.

Reply:
Of course, you would also need the site packages that nixpkgs' `buildPythonPackage` creates, but I'm sure it's possible to do with a bit of work.

At worst you could probably do some patching of plover and the plugin manager here or there to enable it.

Reply:
Upon research you may be interested in the `PYTHONUSERBASE` environmental variable, although I'm not sure if Nixpkgs already co-opts it or not.